Sources/PriorityQueueModule/PriorityQueue.swift

```diff
@@ -0,0 +1,363 @@
+
+                let parentIdx = _parentIndex(of: largestDescendantIdx)
+                if storage[largestDescendantIdx] < storage[parentIdx] {
+                    storage.swapAt(largestDescendantIdx, parentIdx)
+                }
+
+                _trickleDownMax(startingAt: largestDescendantIdx)
+            }
+        }
+    }
+
+    /// Returns the smallest or largest child or grandchild of the element at the given index,
+    /// as determined by the predicate.
+    ///
+    /// - parameter index: The index of the item whose descendants should be compared.
+    /// - parameter predicate: Returns `true` if its first argument should be ordered before its second argument.
+    private func _indexOfChildOrGrandchild(
+        of index: Int,
+        sortedUsing predicate: (Element, Element) -> Bool
+    ) -> (index: Int, isChild: Bool)? {
+        let leftChildIdx = _leftChildIndex(of: index)
+        guard leftChildIdx < storage.count else {
+            return nil
+        }
+
+        var result: (index: Int, isChild: Bool) = (leftChildIdx, true)
+
+        let rightChildIdx = _rightChildIndex(of: index)
+        guard rightChildIdx < storage.count else {
+            return result
+        }
+
+        // Compare the two children
+        if predicate(storage[rightChildIdx], storage[leftChildIdx]) {
+            result.index = rightChildIdx
+        }
+
+        let firstGrandchildIdx = _firstGrandchildIndex(of: index)
+        guard firstGrandchildIdx < storage.count else {
+            return result
+        }
+
+        // Iterate through the grandchildren
+        for i in firstGrandchildIdx..._lastGrandchildIndex(of: index) {
+            guard i < storage.count else {
+                return result
+            }
+
+            if predicate(storage[i], storage[result.index]) {
+                result.index = i
+                result.isChild = false
+            }
+        }
+
+        return result
+    }
+
+    // MARK: - Helpers
+
+    @inline(__always)
+    private func _parentIndex(of index: Int) -> Int {
+        (index - 1) / 2
+    }
+
+    @inline(__always)
+    private func _grandparentIndex(of index: Int) -> Int {
+        (index - 3) / 4
+    }
+
+    @inline(__always)
+    private func _leftChildIndex(of index: Int) -> Int {
+        index * 2 + 1
+    }
+
+    @inline(__always)
+    private func _rightChildIndex(of index: Int) -> Int {
+        index * 2 + 2
+    }
+
+    @inline(__always)
+    private func _firstGrandchildIndex(of index: Int) -> Int {
+        index * 4 + 3
+    }
+
+    @inline(__always)
+    private func _lastGrandchildIndex(of index: Int) -> Int {
+        index * 4 + 6
+    }
+}
+
+/// Returns `true` if `count` elements falls on a min level in a min-max heap.
+///
+/// - Precondition: `count` must be > 0.
+@inline(__always)
+func _minMaxHeapIsMinLevel(_ count: Int) -> Bool {
+    precondition(count > 0)
+
+    return count._binaryLogarithm() & 0b1 == 0
+}
```

Comment:
I had moved this out in my original implementation that was generic around both the element and its comparable key path, because it seemed strange to have to specify the generics in the tests.